src/lib/homestore_backend/replication_state_machine.cpp

```diff
@@ -46,8 +46,50 @@ void ReplicationStateMachine::on_rollback(int64_t lsn, sisl::blob const&, sisl::
 
 void ReplicationStateMachine::on_error(ReplServiceError error, const sisl::blob& header, const sisl::blob& key,
                                        cintrusive< repl_req_ctx >& ctx) {
-    LOGE("on_error  with :{}, lsn {}", error, ctx->lsn);
-    // TODO:: block is already freeed at homestore side, handle error if necessay.
+
+    RELEASE_ASSERT(ctx->is_proposer, "on_error should only be called from proposer");
+    const ReplicationMessageHeader* msg_header = r_cast< const ReplicationMessageHeader* >(header.cbytes());
+    LOGE("on_error, message type {} with lsn {}, error {}", msg_header->msg_type, ctx->lsn, error);
+    switch (msg_header->msg_type) {
+    case ReplicationMessageType::CREATE_PG_MSG: {
+        if (ctx) {
+            auto result_ctx = boost::static_pointer_cast< repl_result_ctx< PGManager::NullResult > >(ctx).get();
+            result_ctx->promise_.setValue(folly::makeUnexpected(homeobject::toPgError(error)));
+        }
+        break;
+    }
+    case ReplicationMessageType::CREATE_SHARD_MSG:
+    case ReplicationMessageType::SEAL_SHARD_MSG: {
+        if (ctx) {
+            auto result_ctx =
+                boost::static_pointer_cast< repl_result_ctx< ShardManager::Result< ShardInfo > > >(ctx).get();
+            result_ctx->promise_.setValue(folly::makeUnexpected(toShardError(error)));
+        }
+        break;
+    }
+
+    case ReplicationMessageType::PUT_BLOB_MSG: {
+        if (ctx) {
+            auto result_ctx =
+                boost::static_pointer_cast< repl_result_ctx< BlobManager::Result< HSHomeObject::BlobInfo > > >(ctx)
+                    .get();
+            result_ctx->promise_.setValue(folly::makeUnexpected(toBlobError(error)));
+        }
+        break;
+    }
+    case ReplicationMessageType::DEL_BLOB_MSG:
+        if (ctx) {
+            auto result_ctx =
+                boost::static_pointer_cast< repl_result_ctx< BlobManager::Result< HSHomeObject::BlobInfo > > >(ctx)
+                    .get();
+            result_ctx->promise_.setValue(folly::makeUnexpected(toBlobError(error)));
+        }
+        break;
+    default: {
+        LOGE("Unknown message type, error unhandled , error :{}, lsn {}", error, ctx->lsn);
```

Comment:
can we add a `UnknowMessageTypeError` for this case?
if we do nothing here, it will still block, right?  because `result_ctx->promise_.setValue` is not called

Reply:
for unknown message you dont know the type of the result_ctx, not to say setValue

Reply:
IMO , if we can not handle this,  it is better to terminate the process here rather than leaving a stuck one , since the crash incident can be captured by our monitor or something else , which can at least give us an alert to look into what happens.